### ISSUE TITLE
`Nothing` instead of `error`?

### ISSUE THREAD

**Issue by TomMD** [GaloisInc/pure-zlib#3]
Is there a reason to use `error` in `decompress` instead of `Nothing`?  It seems `Nothing` is suitable and even preferred in many cases.  We could optionally also provide `decompressOrFail` and return `Either String` in the style of `binary`.  If this works I can submit a patch.

**Comment by acw** [comment]
Laziness, basically. A version that returned an Either would probably be more appropriate, as we could return some form of error message.

**Comment by vandenoever** [comment]
I cannot use pure-zlib at the moment because decompress can give an error. I need a zlib implementation that can be used in pure code. The `decompress` signature is
```haskell
decompress :: ByteString -> Maybe ByteString
```
but the error is still possible from many places in the code. I would expect any error to result in a `Nothing` return value, not a runtime error.

**Comment by acw** [comment]
OK, good to know. Can I assume that:

```
decompress :: ByteString -> Either DecompressError ByteString
```

or the always popular:

```
decompress :: ByteString -> Either String ByteString
```

would work for you, as well?

**Comment by vandenoever** [comment]
They would.
Another issue is that currently `pure-zlib` clashes with normal `zlib`. Both are in `Codec.Compression.Zlib`. My code needs compression and decompression. GHC has [an extension](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#package-qualified-imports) that makes it possible to use both anyway.
```haskell
import "zlib"      Codec.Compression.Zlib (compress)
import "pure-zlib" Codec.Compression.Zlib (decompress)
```

**Comment by acw** [comment]
OK. I actually clashed on purpose, to try to minimize the changes required switching between the two. Hmmmm.

**Comment by vandenoever** [comment]
That would work if all the signatures were the same and the implementation is complete. Until it is, it is more convenient to have it in a different package. Once the implementations is complete alternative you might add a package that wraps the pure package in a zlib compatible package.

**Comment by acw** [comment]
Well, returning to the original bug -- and please consider filing another bug regarding the signatures -- this should now be fixed in HEAD. `decompress` now returns an `Either` containing either a `DecompressionError` (which is, happily, a member of `Exception`, should the obvious function be of use) or the resulting `ByteString`. This task turned out a bit more invasive and complicated than I had planned, with a side trip through some additional testing infrastructure.

I have marked this work as version 0.6; travis feels that it builds on 7.6 and above; and I have uploaded the result to Hackage.

**Comment by vandenoever** [comment]
https://github.com/GaloisInc/pure-zlib/commit/ec233ce8eb534e44d280fb7eb3657d1952ef761f is indeed a large patch. The code still contains `raise` though.

Errors and exceptions cannot be caught in pure code.

**Comment by TomMD** [comment]
`raise` is from `ExceptionM` - it is not an IO exception.

**Comment by vandenoever** [comment]
@TomMD how can it be caught in pure code? Could you give an example where you raise and catch the exception in pure code?

**Comment by acw** [comment]
It does not make it out to your code; it pops out as the result of runDeflateM.

**Comment by TomMD** [comment]
@vandenoever 

```
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}
import MonadLib (ExceptionM, raise, runM)
import MonadLib.Monads (ExceptionT,Id)

main :: IO ()
main =
 do print (runM f :: Either String Int)
    print (case runM f of
            Left _  -> "Look, I 'caught' an 'exception'"
            Right r -> show r)

f :: ExceptionT String Id Int
f = raise "message"
```

**Comment by vandenoever** [comment]
@TomMD thank you for the explanation. I was not aware that this was possible.

**Comment by acw** [comment]
Yes, thanks @TomMD. @vandenoever, welcome to the joy of Haskell and Monad stacks :smile:.
